base/iterator.jl

```diff
@@ -478,9 +478,14 @@ Put differently, the elements of the argument iterator are concatenated. Example
 flatten(itr) = Flatten(itr)
 
 eltype{I}(::Type{Flatten{I}}) = eltype(eltype(I))
-iteratorsize{I}(::Type{Flatten{I}}) = SizeUnknown()
 iteratoreltype{I}(::Type{Flatten{I}}) = iteratoreltype(eltype(I))
 
+flatten_iteratorsize{T<:Tuple}(::Union{HasShape, HasLength}, b::Type{T}) = HasLength()
+flatten_iteratorsize(a, b) = SizeUnknown()
+flatten_length{T<:Tuple}(f, ::Type{T}) = isleaftype(T) ? nfields(T)*length(f.it) : ArgumentError("Cannot compute length of a tuple-type which is not a leaf-type")
```

Comment:
should this actually throw the ArgumentError instead of returning it? and test that branch?